crates/runtime-datafusion-udfs/src/embed.rs

```diff
@@ -0,0 +1,511 @@
+    logical_expr::{ColumnarValue, ScalarUDFImpl, Signature, TypeSignature, Volatility},
+};
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::{Arc, LazyLock};
+use tokio::sync::RwLock;
+
+pub static EMBED_UDF_NAME: &str = "embed";
+pub static DOCUMENTATION: LazyLock<Documentation> = LazyLock::new(|| Documentation {
+    doc_section: DocSection::default(),
+    description: "Generates embeddings for text using a specified embedding model".to_string(),
+    syntax_example: "embed(text, model_name)".to_string(),
+    sql_example: Some("SELECT embed('hello world', 'potion_2m')".to_string()),
+    arguments: Some(vec![
+        ("text".to_string(), "The text string to embed.".to_string()),
+        (
+            "model_name".to_string(),
+            "The name of the embedding model to use as defined in the Spicepod.".to_string(),
+        ),
+    ]),
+    alternative_syntax: Some(vec!["embed(['foo', 'bar'], 'potion_2m')".to_string()]),
+    related_udfs: None,
+});
+
+pub static SIGNATURE: LazyLock<Signature> = LazyLock::new(|| {
+    Signature::one_of(
+        // In order of least likely to auto-coerce, via logical_expr docs for OneOf:
+        // > Coercion is attempted to match the signatures in order,
+        // > and stops after the first success, if any.
+        vec![
+            // embed(make_array(a, b, c), model_name)
+            TypeSignature::Exact(vec![
+                DataType::List(Arc::new(Field::new("sentence", DataType::Utf8, true))),
+                DataType::Utf8,
+            ]),
+            // embed(text, model_name)
+            TypeSignature::Exact(vec![DataType::Utf8, DataType::Utf8]),
+        ],
+        Volatility::Stable,
+    )
+});
+
+pub type EmbeddingModelStore = HashMap<String, Arc<dyn llms::embeddings::Embed>>;
+
+macro_rules! string_array_iter {
+    ($array:expr) => {{
+        let iter: Box<dyn Iterator<Item = Option<&str>>> = match $array.data_type() {
+            &DataType::Utf8 => Box::new(as_string_array($array)?.iter()),
+            &DataType::LargeUtf8 => Box::new(as_large_string_array($array)?.iter()),
+            other_data_type => return exec_err!("Expected strings, got {other_data_type:?}"),
+        };
+
+        iter
+    }};
+}
+
+#[derive(Debug)]
+pub struct Embed {
+    model_store: Arc<RwLock<EmbeddingModelStore>>,
+}
+
+impl Embed {
+    #[must_use]
+    pub fn new(model_store: Arc<RwLock<EmbeddingModelStore>>) -> Self {
+        Self { model_store }
+    }
+
+    fn embed_single(
+        model: &Arc<dyn llms::embeddings::Embed>,
+        sentence: &str,
+    ) -> DataFusionResult<ColumnarValue> {
+        let embedding = model
+            .embed_sync(EmbeddingInput::String(sentence.to_owned()))
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+        let vector_size = match embedding.first() {
+            Some(embedding) => embedding.len(),
+            _ => unreachable!("Should have at least one embedding"),
```

Comment:
Just to confirm, this should be truly unreachable, correct? Else we should avoid the panic.

Reply:
Correct